neutrinordp/xrdp-neutrinordp.c

```diff
@@ -94,10 +94,107 @@ lxrdp_start(struct mod *mod, int w, int h, int bpp)
 
 /******************************************************************************/
 /* return error */
+static void
+set_keyboard_overrides(struct mod *mod)
+{
+    const struct kbd_overrides *ko = &mod->kbd_overrides;
+    rdpSettings *settings  = settings = mod->inst->settings;
+
+    if (mod->allow_client_kbd_settings)
+    {
+        settings->kbd_type = mod->client_info.keyboard_type;
+        settings->kbd_subtype = mod->client_info.keyboard_subtype;
+        /* Define the most common number of function keys, 12.
+           because we can't get it from client. */
+        settings->kbd_fn_keys = 12;
+
+        settings->kbd_layout = mod->client_info.keylayout;
+        if (mod->client_info.keyboard_type == 0x00)
+        {
+            LOG(LOG_LEVEL_WARNING, "keyboard_type:[0x%02x] ,Set on Server",
+                mod->client_info.keyboard_type  );
+        }
+        else if (mod->client_info.keyboard_type == 0x04)
+        {
+            /* When the keyboard type is 0x04 (PC101/102 keyboard) and the lower 16
+               bits are a value of 0x0411(Japanese KeyLayout), the kbd_layout is set
+               to 0x00000411 no matter what the upper 16 bits are. */
+            if ((settings->kbd_layout & 0x0000FFFF) == 0x00000411)
+            {
+                settings->kbd_layout = 0x00000411;
+            }
+            LOG(LOG_LEVEL_DEBUG,
+                "keyboardLayout for PC101/102 keyboard:"
+                "[0x%08X]", settings->kbd_layout);
+        }
+        else if (mod->client_info.keyboard_type == 0x07)
+        {
+            /* When the keyboard type is 0x07 (Japanese keyboard) and the lower 16
+               bits are a value of 0x0411(Japanese KeyLayout), the kbd_layout is set
+               to 0x00000411 no matter what the upper 16 bits are. */
+            if ((settings->kbd_layout & 0x0000FFFF) == 0x00000411)
+            {
+                settings->kbd_layout = 0x00000411;
+            }
```

Comment:
I realized that `kbd_type 0x07` could also result in a `keyboardLayout` other than `0x????0411`.

So I decided not to simply set `kbd_layout = 0x00000411` for `kbd_type 0x07`.
Instead, I set `0x00000411` only when the lower 16 bits of `keyboardLayout` are `0x0411`.

Reply:
Hi @TOMATO-ONE 

Thanks for those changes.

I don't feel qualified to comment on the Japanese keyboard changes. Not only do I not have suitable keyboard(s) available, but I wouldn't be able to use them even if I did.

@metalefty - how does this look to you?

Reply:
@metalefty  I would like to ask you to do the same.
The `keyboardLayout`=`0xnnnn0411` at `kbd_Type`=`0x04` causes a key layout mismatch problem in xrdp itself.
I would like to address only the NeutrinoRDP Proxy module issue in this PR, and consider the xrdp side in another PR.